Initial docs for new adapter testing framework (#1263)
fishtown-analytics/docs.getdbt.com: website/docs/docs/contributing/testing-a-new-adapter.md

---
title: "Testing a new adapter"
id: "testing-a-new-adapter"
---

<VersionBlock lastVersion="1.0">

The previous way to test dbt adapter functionality was via a packaged suite of tests, [`pytest-dbt-adapter`](https://github.com/dbt-labs/dbt-adapter-tests). We've since deprecated that approach in favor of the new testing framework outlined in this document.

> </VersionBlock>

Author: Is this callback valuable? Or should we pretend like it never existed?

When we're ready, we will want to officially "deprecate" the `dbt-adapter-tests` repo by adding a note to its README, and then archive it. Similar to what we did for https://github.com/dbt-labs/dbt-integration-tests. (We're running out of names!)

Reviewer: I think we might want the deprecation mentioned for a while. We wouldn't want to confuse somebody by its sudden disappearance.

dbt-core offers a standard framework for running pre-built functional tests, and defining new ones.

The core testing framework is built using `pytest`, a mature and standard library for testing Python projects. It assumes a basic level of familiarity with the foundational concepts in `pytest`, such as fixtures. If this is your first time using `pytest`, we recommend you read the documentation: https://docs.pytest.org/

> The [`tests` module](https://github.com/dbt-labs/dbt-core/tree/HEAD/core/dbt/tests) within `dbt-core` includes basic utilities for setting up `pytest` + `dbt`. These are built into `dbt-core`, and used by all functional tests. These also make it possible to write your own tests, using the same utilities.

Author: Realizing that the module contents here will actually be an dbt-core feature, for now undocumented, that is available to all project/package maintainers as well (if they care to write `pytest` cases)

Let's think about where the docs for this piece should live long-term, given that the intended audience is a bit wider than just creators/maintainers of adapter plugin

Reviewer: We probably want some high level documentation of how to use the standard fixtures and structure a simple test.

```bash
pip install "git+https://github.com/dbt-labs/dbt-core.git#egg=dbt-tests-adapter&subdirectory=tests/adapter"
```

Or specify it in a requirements file like:
<File name="dev_requirements.txt">

```txt
git+https://github.com/dbt-labs/dbt-core.git#egg=dbt-tests-adapter&subdirectory=tests/adapter
```
</File>

### Set up and configure pytest

First, set yourself up to run `pytest` by creating a file named `pytest.ini` at the root of your repository:

<File name="pytest.ini">

```python
[pytest]
filterwarnings =
    ignore:.*'soft_unicode' has been renamed to 'soft_str'*:DeprecationWarning
    ignore:unclosed file .*:ResourceWarning
env_files =
    test.env  # this allows you to use env vars from a file named test.env
              # be sure to add "test.env" to .gitignore as well!
testpaths =
    tests/functional  # name per convention
```

</File>

Then, create a configuration file within your tests directory. In it, you'll want to define all necessary profile configuration for connecting to your data platform in local development and continuous integration. We recommend setting these values with environment variables, since this file will be checked into version control.

> <File name="pytest.ini">

Reviewer: This is supposed to be tests/conftest.py

```python
import pytest
import os

# Import the standard functional fixtures as a plugin
# Note: fixtures with session scope need to be local
pytest_plugins = ["dbt.tests.fixtures.project"]

# The profile dictionary, used to write out profiles.yml
@pytest.fixture(scope="class")
def dbt_profile_target(unique_schema):
    return {
        'type': '<myadapter>',
        'threads': 1,
        'host': os.getenv('HOST_ENV_VAR_NAME'),
        'user': os.getenv('USER_ENV_VAR_NAME'),
        ...

>         'schema': unique_schema  # this will be passed in by the testing framework

Reviewer: In the Redshift test I specified the unique_schema, but in at least some of the other adapters I didn't. I'm also setting it in the 'dbt_profile_data' fixture, so it's not actually necessary. I'm thinking we should standardize on not supplying it here.

---
title: "Testing a new adapter"
id: "testing-a-new-adapter"
---

:::info

> Previously, we offered a packaged suite of tests for dbt adapter functionality: [`pytest-dbt-adapter`](https://github.com/dbt-labs/dbt-adapter-tests). We are deprecating that suite, in favor of the newer testing framework outlined in this document. The rest of this document assumes that your plugin is compatible with dbt-core v1.1 or newer.

Reviewer 2: So people don't miss this, let's more to a prerequisites section.
```suggestion
Previously, we offered a packaged suite of tests for dbt adapter functionality: [`pytest-dbt-adapter`](https://github.com/dbt-labs/dbt-adapter-tests). We are deprecating that suite, in favor of the newer testing framework outlined in this document. 
```

:::

> ## Foundation

Reviewer 2: ```suggestion
## About testing adapters
```

Author: I'm being a bit coy here. This section of the docs _could_ be relevant to anyone who wants to use the testing framework, including package/project maintainers who want the ability to write `pytest` cases against their package/project. I'm going to retitle this section `About the testing framework`.

Reviewer 2: Ah I love that title!

> dbt-core offers a standard framework for running pre-built functional tests, and for defining new ones.

Reviewer 2: Changing this to be as specific as we can to help non-native English speakers follow along.
```suggestion
dbt-core offers a standard framework for running pre-built functional tests, and for defining your own tests.
```

> The core testing framework is built using `pytest`, a mature and standard library for testing Python projects. It assumes a basic level of familiarity with the foundational concepts in `pytest`, such as fixtures. If this is your first time using `pytest`, we recommend you read the documentation: https://docs.pytest.org/

Reviewer 2: ```suggestion
The core testing framework is built using `pytest`, a mature and standard library for testing Python projects. It assumes a basic level of familiarity with the foundational concepts in `pytest`, such as fixtures. 
```

The **[`tests` module](https://github.com/dbt-labs/dbt-core/tree/HEAD/core/dbt/tests)** within `dbt-core` includes basic utilities for setting up `pytest` + `dbt`. These are used by all "pre-built" functional tests, and make it possible to quickly write your own tests.

> 

Reviewer 2: Are there any other prereqs to call out?
```suggestion
## Prerequisites
* Make sure your adapter is compatible with dbt-core v1.1 or newer.
* Become familiar with `pytest` by reading their documentation: https://docs.pytest.org/
```

Author: good callout! I'll do some light reorganization, and include more signposting at the top

> ### Example

Reviewer 2: What do you think about moving the example to the end and after the Using basic tests (was Getting started: basic tests) section? I feel like more of an intro is helpful context (even for the most baic term "basic test") and those who know all about it will skip ahead. 
```suggestion
## Example using a basic dbt test project
```

Author: Fair, the ordering here is a bit odd; I'm intentionally sowing a seed here for a separate docs page that could exist in the near future. The first section (including this example) could be relevant to programmatic testing of _any_ dbt-based product/project, in a way that's actually broader than the narrower adapter testing use case we're outlining in this page. But, that is our first primary use case, which is why it's all been rolled up into this one page for now.

I'll include some more framing language to make the intention with this section clearer

> In this example, we'll set up a basic dbt project using `pytest` fixtures, run some simple commands, and ensure they succeed (or fail) as we expect. To that end, we'll be using the `run_dbt` utility defined within the `tests` module of `dbt-core`.

Reviewer 2: ```suggestion
This example uses the `run_dbt` utility defined within the `tests` module of `dbt-core` to:
* Set up a basic dbt test project using `pytest` fixtures.
* Run some simple commands.
* Ensure they succeed (or fail) as we expect. 
```

> (For the sake of simplicity, the example assumes that we've already installed and configured `pytest`. We'll show that setup in more detail below.)

Reviewer 2: ```suggestion
This example assumes that you've already installed and configured `pytest`. You can see how to do this in [Getting started running basic tests](#getting-started-running-basic-tests)
```

> First, let's define our project, using Python strings to stand in for file contents. The appeal of defining these in a separate file is that we can reuse the same "project" across test cases.

Reviewer 2: It might be helpful to number these steps. I also like starting with a clear action and adding our warmth and helpfulness in the rest of the sentence.
```suggestion
1. Define the project using Python strings to stand in for file contents. Defining these in a separate file will help you reuse the same "project" across different test cases.
```

<File name="tests/functional/example/fixtures.py">

```python
# seeds/my_seed.csv
my_seed_csv = """
id,name,some_date
1,Easton,1981-05-20T06:46:51
2,Lillian,1978-09-03T18:10:33
3,Jeremiah,1982-03-11T03:59:51
4,Nolan,1976-05-06T20:21:35
""".lstrip()

# models/my_model.sql
my_model_sql = """
select * from {{ ref('my_seed') }}
union all
select null as id, null as name, null as some_date
"""

# models/my_model.yml
my_model_yml = """
version: 2
models:
  - name: my_model
    columns:
      - name: id
        tests:
          - unique
          - not_null  # will not pass
"""
```

</File>

> Next, we'll use these "fixtures" to define our test cases. Following the default `pytest` configurations:

Reviewer 2: ```suggestion
2. Use the "fixtures" to define test cases, following the default `pytest` configurations:
```

    # everything that goes in the "seeds" directory
    @pytest.fixture(scope="class")
    def seeds(self):
        return {
            "my_seed.csv": my_seed_csv,
        }

    # everything that goes in the "models" directory
    @pytest.fixture(scope="class")
    def models(self):
        return {
            "my_model.sql": my_model_sql,
            "my_model.yml": my_model_yml,
        }
    
    # the method defines the actual sequence of steps to execute
    # pytest will take care of all "setup" + "teardown"
    def test_run_seed_test(self, project):
        """Seed, then run, then test. We expect one of the tests to fail
           An alternative pattern is to use pytest "xfail" (see below)
        """
        # seed seeds
        results = run_dbt(["seed"])
        assert len(results) == 1
        # run models
        results = run_dbt(["run"])
        assert len(results) == 1
        # test tests
        results = run_dbt(["test"], expect_pass = False) # expect failing test
        assert len(results) == 2

    @pytest.mark.xfail
    def test_build(self, project):
        """Expect a failing test"""
        # do it all
        results = run_dbt(["build"])
```

> Then, we invoke `pytest`:

Reviewer 2: ```suggestion
3. Type the following line of code to invoke `pytest`.  
```

```sh
$ python3 -m pytest tests/functional/test_example.py
=========================== test session starts ============================
platform ... -- Python ..., pytest-..., pluggy-...
rootdir: ...
plugins: ...

tests/functional/test_example.py .X                                  [100%]

======================= 1 passed, 1 xpassed in 1.38s =======================
```

</File>

> The [pytest usage docs](https://docs.pytest.org/how-to/usage.html) offer guidance and a full command reference. In our experience, it can be particularly helpful to use the `-s` flag (or `--capture=no`) to print logs from the underlying dbt invocations, and to step into an interactive debugger if you've added one. You can also use environment variables to set [global dbt configs](global-configs), such as `DBT_DEBUG` (to show debug-level logs).

Reviewer 2: ```suggestion
You can find more ways to run tests, along with a full command reference, in the [pytest usage docs](https://docs.pytest.org/how-to/usage.html). 

We've found the `-s` flag (or `--capture=no`) helpful to print logs from the underlying dbt invocations, and to step into an interactive debugger if you've added one. You can also use environment variables to set [global dbt configs](global-configs), such as `DBT_DEBUG` (to show debug-level logs).

```

> ## Testing your adapter

Reviewer 2: I'm flexible on this title just trying to capture what this section involves and it's more background info than steps for testing your adapter.
```suggestion
## Understanding adapter testing
```

> The framework presented above is available for use by anyone who installs `dbt-core`, and who wishes to define their own test cases.

Reviewer 2: ```suggestion
Anyone who installs `dbt-core`, and wishes to define their own test cases can use the framework presented in the [Example using a basic dbt test project](#example-using-a-basic-dbt-test-project).
```

> Building on top of that foundation, we have also built and made available a [package of reusable adapter test cases](https://github.com/dbt-labs/dbt-core/tree/HEAD/tests/adapter), for creators and maintainers of adapter plugins. These test cases cover basic expected functionality, as well as functionality that frequently requires different implementations across databases.

Reviewer 2: ```suggestion
We've also built and made available a [package of reusable adapter test cases](https://github.com/dbt-labs/dbt-core/tree/HEAD/tests/adapter), for creators and maintainers of adapter plugins. These test cases cover basic expected functionality, as well as functionality that frequently requires different implementations across databases.
```

2. **Optional tests**, for second-order functionality that is common across plugins, but not required for basic use. Your plugin can opt into these test cases by inheriting existing ones, or reimplementing them with adjustments. More tests will be added as we convert older tests defined on dbt-core and mature plugins to use the standard framework.

3. **Custom tests**, for behavior that is specific to your adapter / data platform. Each data warehouse has its own specialties and idiosyncracies. We encourage you to use the same `pytest`-based framework, utilities, and fixtures to write your own custom tests for functionality that is unique to your adapter.

If you run into an issue with the core framework, or the basic/optional test cases—or if you've written a custom test that you believe would be relevant and useful for other adapter plugin developers—please open an issue or PR in the `dbt-core` repository on GitHub.

> ## Getting started: basic tests

Reviewer 2: ```suggestion
## Getting started running basic tests
```

In this section, we'll walk through the three steps to start running our basic test cases on your adapter plugin:

1. Install dependencies
2. Set up and configure pytest
3. Define test cases

> ### Install dependencies

Reviewer 2: Maybe we can come back later and number the steps in these procedures. I'm leaving them as-is for now.

:::

This document has two sections:
1. ["About the testing framework"](#about-the-testing-framework) describes the standard framework that we maintain for using pytest together with dbt. It includes an example that shows the anatomy of a simple test case.
2. ["Testing your adapter"](#testing-your-adapter) offers a step-by-step guide for using our out-of-the-box suite of "basic" tests, which will validate that your adapter meets a baseline of dbt functionality.

## Prerequisites
- Your adapter must be compatible with dbt-core **v1.1** or newer
- You should be familiar with **pytest**: https://docs.pytest.org/

## About the testing framework

dbt-core offers a standard framework for running pre-built functional tests, and for defining your own tests. The core testing framework is built using `pytest`, a mature and standard library for testing Python projects.

The **[`tests` module](https://github.com/dbt-labs/dbt-core/tree/HEAD/core/dbt/tests)** within `dbt-core` includes basic utilities for setting up pytest + dbt. These are used by all "pre-built" functional tests, and make it possible to quickly write your own tests.

Those utilities allow you to do three basic things:

> 1. **Quickly set up a dbt "project."** Define project resources via methods such as `models()` and `seeds()`. Use `project_config_update()` to pass configurations into `dbt_project.yml`.

Reviewer 3: Just synced with @VersusFacit on something related to starting using this testing framework. It would be great if we can highlight two things: 
- the concept of pytest fixtures and fixture scope is important when writing tests
- we choose the `class` scope for most of our fixtures.

2. **Define a sequence of dbt commands.** The most important utility  is `run_dbt()`, which returns the [results](dbt-classes#result-objects) of each dbt command. It takes a list of CLI specifiers (subcommand + flags), as well as an optional second argument, `expect_pass=False`, for cases where you expect the command to fail.

> 3. **Validate the results of those dbt commands.** For example, `check_relations_equal()` asserts that two database objects have the same structure and content. You can also write your own `assert` statements, by inspecting the results of a dbt command, or querying arbitrary database objects with `project.run_sql()`.

Reviewer 3: Should we consider add a link [to this file](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/tests/util.py) so folks can find out the latest functions we provide?
